tests/src/test/java/alluxio/BaseIntegrationTest.java

```diff
@@ -12,12 +12,52 @@
 package alluxio;
 
 import org.junit.Rule;
+import org.junit.rules.TestWatcher;
 import org.junit.rules.Timeout;
+import org.junit.runner.Description;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
 
 /**
  * Base class used for specifying the maximum time a test should run for.
  */
 public abstract class BaseIntegrationTest {
+  private static final File LOG_DIR = new File("./target/logs");
+  private static final File TESTS_LOG = new File("./target/logs/tests.log");
+
   @Rule
   public Timeout mGlobalTimeout = Timeout.millis(Constants.MAX_TEST_DURATION_MS);
+
+  @Rule
+  public TestWatcher mWatcher = new TestWatcher() {
+    // When tests fail, save the logs.
+    protected void failed(Throwable e, Description description) {
+      try {
+        Files.copy(Paths.get(TESTS_LOG.toURI()),
+            Paths.get(LOG_DIR.getAbsolutePath(), String.format("%s-%s.log",
+                description.getClassName(), description.getMethodName())),
+            StandardCopyOption.REPLACE_EXISTING);
+      } catch (IOException e1) {
+        if (e != null) {
+          e.addSuppressed(e1);
+        } else {
+          throw new RuntimeException(e1);
+        }
+      }
+      return;
+    }
+
+    protected void succeeded(Description description) {
+      // When tests succeed, reset the log file.
+      try {
+        Files.write(Paths.get(TESTS_LOG.getAbsolutePath()), "".getBytes());
```

Comment:
Is there a nicer way to do this with something like `truncate`?

Reply:
I don't see a `truncate` in any libraries we use

Reply:
ok I found a nicer way with `new FileWriter(Constants.TESTS_LOG).close();`